apps/src/lib/util/speech.js

```diff
@@ -22,38 +20,153 @@ export function textToSpeech(
   region,
   appLanguages
 ) {
-  const speechConfig = SpeechConfig.fromAuthorizationToken(token, region);
-  speechConfig.speechSynthesisOutputFormat =
-    SpeechSynthesisOutputFormat.Audio16Khz32KBitRateMonoMp3;
-  let voice =
-    (appLanguages[language] && appLanguages[language][gender]) ||
-    appLanguages['English']['female'];
-  const synthesizer = new SpeechSynthesizer(
-    speechConfig,
-    undefined /* AudioConfig */
-  );
-  let ssml = `<speak version="1.0" xmlns="https://www.w3.org/2001/10/synthesis" xml:lang="en-US"><voice name="${voice}">${text}</voice></speak>`;
-  synthesizer.speakSsmlAsync(
-    ssml,
-    result => {
-      // There is no way to make ajax requests from html on the filesystem.  So
-      // the only way to play sounds is using HTML5. This scenario happens when
-      // students export their apps and run them offline. At this point, their
-      // uploaded sound files are exported as well, which means varnish is not
-      // an issue.
-      const forceHTML5 = window.location.protocol === 'file:';
-      Sounds.getSingleton().playBytes(result.audioData, {
+  if (text.length > 749) {
+    text = text.slice(0, 750);
+    outputWarning(i18n.textToSpeechTruncation());
+  }
+  let voice = appLanguages['English']['female'];
+  let languageCode = 'en-US';
+  if (appLanguages[language] && appLanguages[language][gender]) {
```

Comment:
If we have some gender of the chosen language, but not the requested language, should we fall back on a different gender of the chosen language rather than English Female?

Reply:
technically that shouldn't happen because of the way we're getting the language options here: https://github.com/code-dot-org/code-dot-org/pull/35571/files#diff-1f68e16ab551449bdbd0fb7f01004f68 (we only offer the language if we have male and female voices available). This is more to check that they chose something from the dropdowns (didn't misspell female or English)